syncronous import via can-import tags

nlundquist opened the pull request:
Passes `can-import`ed modules to stache in helpers scope. 
Add test to check if can-import tags immediately have `export` property on view model when loaded by steal-stache.

Part of the resolution of canjs/can-view-import#111.

Test on this will fail until changes in canjs/can-view-import#124 are available.

nlundquist committed:
pass `can-import`ed modules to stache in helpers scope so they can be available synchronously

nlundquist committed:
fix rebase

nlundquist committed:
convert to tabs

nlundquist commented on steal-stache.js:
```diff
@@ -5,21 +5,30 @@ var loader = require("@loader");
+		"\t\t\tmoduleOptions.helpers = assign({ module: module, tagImportMap: tagImportMap }, moduleOptions.helpers);\n" +
 		"\t\t} else {\n" +
 		"\t\t\tmoduleOptions.module = module;\n" +
+		"\t\t\tmoduleOptions.tagImportMap = tagImportMap;\n" +
```
this may be unnecessary? wasn't clear to me in what cases the `helper` object would be missing

justinbmeyer replied:
yeah, why is this being done?

nlundquist replied:
removed this in favor of always passing `module` as part of the LetContext. If someone knows why it needed to be done this way we can add it back.

nlundquist commented on steal-stache.js:
```diff
@@ -5,21 +5,30 @@ var loader = require("@loader");
 	intermediate = JSON.stringify(intermediate);
 
-	return "define("+imports+",function(module, assign, stache, mustacheCore){ \n" +
+	return "define("+imports+",function(module, assign, stache, mustacheCore, canViewImport, canStacheBindings){ \n" +
```
don't _need_ to add these two new arguments, but it's nice to make it clear what 6 arguments we slice off below

justinbmeyer replied:
plus a minifier should be able to shorten this.

justinbmeyer commented on steal-stache.js:
```diff
@@ -5,21 +5,30 @@ var loader = require("@loader");
+		"\tvar tagImports = Array.prototype.slice.call(arguments, 6);\n" +
 		"\treturn function(scope, options, nodeList){\n" +
 		"\t\tvar moduleOptions = assign({}, options);\n" +
+		"\t\tvar tagImportMap = " + tagImportNames + ".reduce((map, name, index) => {\n" +
```
is reduce in IE11?  Are arrow functions?  This should be ES5.

matthewp replied:
reduce is, arrows are not

justinbmeyer commented on steal-stache.js:
```diff
@@ -5,21 +5,30 @@ var loader = require("@loader");
+		"\n"+
 		"\t\tif(moduleOptions.helpers) {\n" +
-		"\t\t\tmoduleOptions.helpers = assign({ module: module }, moduleOptions.helpers);\n" +
+		"\t\t\tmoduleOptions.helpers = assign({ module: module, tagImportMap: tagImportMap }, moduleOptions.helpers);\n" +
```
this shouldn't be added as a helper, it should be added as a LetContext.  A `LetContext` is only available to the particular template.  cc @phillipskevin 

Arguably in 6.0, we should remove being able to pass helpers as a second argument and have people pass a properly constructed `scope`.  (Maybe do this in 7)

justinbmeyer replied:
https://github.com/canjs/can-view-scope/blob/28b36610a020098aa26ddf641cb8314b7409d9bf/can-view-scope.js#L718

justinbmeyer replied:
A bit of discovery is probably needed here.  I'm not sure everywhere `let contexts` are normally made.  Here's one place:

https://github.com/canjs/can-stache/blob/bf87a5acc05dbb3a38513903807e22b39b1948a5/can-stache.js#L546

I feel like we're going to have to check the `scope` here and see if it's, in fact, an instance of Scope.  

If it is, we might want to add these imports to that scope's let context.

If not, we should create a scope and add our own let context with these imports.

The function listed above does much of the same thing.  If we end up re-creating it too much, we might want to export it from stache and re-use it in can-view-import.

nlundquist replied:
Should we also make it so `module` is passed as part of the let context rather than as a helper? Seems to make sense for the same reasoning as including `tagImportMap` as a let variable rather than a helper.

justinbmeyer replied:
What does module do? Is tagImportMap added directly? If it is, why is it?

justinbmeyer reviewed (changes requested):
Imports should be added as a `let` variable, not a helper.

We need to stick to `ES5`.

nlundquist committed:
change arrow to regular function

nlundquist committed:
use LetContext to pass tagImportMap & module rather than sticking them in the helpers
only create a new Scope if scope isn't one already
only create a new LetContext if one isn't already in the scope

nlundquist commented:
I've updated this to use a LetContext rather than sticking things on the `helpers` object.
Removed the arrow function usage.

nlundquist commented on steal-stache.js:
```diff
@@ -5,21 +5,31 @@ var loader = require("@loader");
+		"\t\t}, {});\n" +
+		"\n"+
+		"\t\tif (!(scope instanceof Scope)) { scope = new Scope(scope); }\n" +
+		"\t\tvar variableScope = scope.getScope(function(s) { return s._meta.variable === true });\n" +
```
Should I change this to only check the "leaf" scope to see if it's a LetContext? This currently scans for the closest LetContext which may not be what we want.

nlundquist replied:
Maybe we should always be adding a LetContext :/

nlundquist committed:
get `module` tests passing now that it's in LetContext

justinbmeyer commented:
@nlundquist can you figure out why CI is breaking and fix it? Thanks!

nlundquist commented:
hey @justinbmeyer the CI isn't running because we haven't released changes this depends on in canjs/can-view-import#124.

This dependency mentioned in the PR description above. This can be tested locally by NPM linking those changes.

Once canjs/can-view-import#124 gets approved I'll update this PR to point at the new release of can-view-import to get it passing.

nlundquist committed:
update test to use property name "module" & confirm that binding is made to a let context

nlundquist committed:
use can-view-import#major
